docs/proposals/sysctl.md

```diff
@@ -0,0 +1,376 @@
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+## Abstract
+
+This proposal aims at extending the current pod specification with support
+for namespaced kernel parameters set for each pod.
+
+In Linux, the sysctl interface allows an administrator to modify kernel
+parameters at runtime. Parameters are available via `/proc/sys/` virtual
+process file system. The parameters cover various subsystems such as:
+
+* kernel (common prefix: `kernel.`)
+* networking (common prefix: `net.`)
+* virtual memory (common prefix: `vm.`)
+* MDADM (common prefix: `dev.`)
+
+More subsystems are described in [Kernel docs](https://www.kernel.org/doc/Documentation/sysctl/README).
+
+To get a list of basic prefixes on your system, you can run
+
+```
+$ sudo sysctl -a | cut -d' ' -f1 | cut -d'.' -f1 | sort -u
+```
+
+To get a list of all parameters, you can run
+
+```
+$ sudo sysctl -a
+```
+
+A number of them are namespaced and can therefore be set for a container
+independently with today's Linux kernels.
+
+**Note**: This proposal - while sharing some use-cases - does not cover ulimits
+(compare [Expose or utilize docker's rlimit support](https://github.com/kubernetes/kubernetes/issues/3595)).
+
+## Motivation
+
+A number of Linux applications need certain kernel parameter settings to
+
+- either run at all
+- or perform well.
+
+In Kubernetes we want to allow to set these parameters within a pod specification
+in order to enable the use of the platform for those applications.
+
+With Docker version 1.11.1 it is possible to change kernel parameters inside privileged containers.
+However, the process is purely manual and the changes might be applied across all containers
+affecting the entire host system. It is not possible to set the parameters within a non-privileged
+container.
+
+With [docker#19265](https://github.com/docker/docker/pull/19265) docker-run as of 1.12.0
+supports setting a number of whitelisted sysctls during the container creation process.
+
+Some real-world examples for the use of sysctls:
+
+- PostgreSQL requires `kernel.shmmax` and `kernel.shmall` (among others) to be
+  set to reasonable high values (compare [PostgresSQL Manual 17.4.1. Shared Memory
+  and Semaphores](http://www.postgresql.org/docs/9.1/static/kernel-resources.html)).
+  The default of 32 MB for shared memory is not reasonable for a database.
+- RabbitMQ proposes a number of sysctl settings to optimize networking: https://www.rabbitmq.com/networking.html.
+- web applications with many concurrent connections require high values for
```

Comment:
Somaxconn is the upper bound on the backlog parameter of the listen syscall. Citing http://serverfault.com/questions/518862/will-increasing-net-core-somaxconn-make-a-difference:

> Sometimes it's preferable to fail fast and let the load-balancer to do it's job(retry) than to make user wait - for that purpose we set net.core.somaxconn any value, and limit application backlog to e.g. 10 and set net.ipv4.tcp_abort_on_overflow to 1.

So there might be side-effects with big somaxconn numbers if applications expect that there is a sensible maximum which is used by the kernel to cap the passed value, like
```
listen(fd, 1000000000)
```
OOM will kill them. If the socket structs in the kernel are lazily allocated, OOM will only kill on high load. Even worse.

In general, it feels like using memory accounting works to be on the safe side regarding pod isolation, BUT it only kicks in when a process actually consumes the resources. On the other hand, sysctl settings will set limits for allocation (= **fail fast**), i.e. that a drawback for relying only on memory accounting.

It might make sense to have a combination of both, accounting and sysctls: the user can set sysctls to increase defaults, but they are bounded by memcg limits.

Side question: does kmem count as "normal" memory in the memcg? Do we plan to add extra limits for kmem? As kmem is locked memory this might make sense.

All this also relates to the suggestion to add awareness to the scheduler: if shmem is accounted by memcg and we set limits on that, the memory expects/limits values in the scheduler are enough to avoid over-commit. It's not completely clear to me yet how global values like kernel.shmmni work though, whether cgroup hierarchies play a role and whether we have a limited resource here if the root cgroup has low values and pods set bigger values.